api/samples/neoncounter.c

```diff
@@ -0,0 +1,703 @@
+
+        if (IS_DEBUG) {
+            fprintf(debug, "%s", op_name);
+        }
+
+        // assure that string is not null
+        if (strlen(op_name) < 1) {
+            continue;
+        }
+
+        // count taken branches
+        if (instr_is_cbr(ins)) {
+            // count all branching instructions
+            counts->branching++;
+            // branching instructions are not simd or load/store instructions, so we
+            // skip the rest
+            if (IS_DEBUG) {
+                fprintf(debug, "\n");
+            }
+            continue;
+        }
+
+        // load instructions
+        if (op_name[0] == 'l') {
+            for (int i = 0; i < 45; i++) {
+                if (op_code == OP_loading[i]) {
+                    is_load = true;
+                    counts->load++;
+                    if (i == 37 || i == 41) { // ld1, ld1r
+                        counts->load_linear++;
+                    } else if ((i > 37 && i < 41) ||
+                               (i > 41 && i < 45)) { // ld2, ld3, ld4, ld2r, ld3r, ld4r
+                        counts->load_structured++;
+                    }
+                    break;
+                }
+            }
+        }
+
+        // store instructions
+        if (op_name[0] == 's') {
+            for (int i = 0; i < 28; i++) {
+                if (op_code == OP_storing[i]) {
+                    is_store = true;
+                    counts->store++;
+                    if (i == 24) { // st1
+                        counts->store_linear++;
+                    } else if (i > 24) { // st2, st3, st4
+                        counts->store_structured++;
+                    }
+                    break;
+                }
+            }
+        }
+
+        // arithmetic instructions
+        for (int i = 0; i < 204; i++) {
+            if (op_code == OP_arithmetic[i]) {
+                is_arith = true;
+                counts->arith++;
+                break;
+            }
+        }
+
+        // check for VECTOR Instructions
+        // -------------------------------------------------- SIMD-Registers: Q0-Q31,
+        // D0-D31, S0-S31, H0-H31, B0-B31 (dr_ir_opnd.h); SVE: Z0-Z31
+        reg_id_t reg;
+
+        // NEON and FP use the same registers
+        // arithmetic NEON Instr. store their width in an additional src-slot
+        // this variable holds the number of src slots that acutally point to a
+        // register
+        int num_src = 0;
+
+        // potential neon load/store instructions should reference a memory block
+        // larger than four doubles (this is an estimate)
+        if (is_load || is_store) {
+            if (instr_memory_reference_size(ins) < sizeof(double) * 4) {
+                if (IS_DEBUG) {
+                    fprintf(debug, "\n");
+                }
+                continue;
+            } else {
+                num_src = instr_num_srcs(ins);
+            }
+        } else {
+            switch (op_code) {
```

Comment:
I think it would be better to move the checks in the loop body to separate functions, e.g. have something like is_neon(), is_vector(), etc. #431 seems loosely related, as it describes a way for DynamoRIO to provide additional info about instructions, e.g. the ISA version.